test/model/topics_test.dart

```diff
@@ -0,0 +1,464 @@
+    Future<void> prepareWithMessages(List<StreamMessage> messages, {
+      required List<Condition<Object?>> expectedTopics,
+    }) async {
+      await prepare(topics: []);
+      assert(messages.isNotEmpty);
+      assert(messages.every((m) => m.streamId == channel.streamId));
+      await store.addMessages(messages);
+      check(model).getChannelTopics(channel.streamId).isNotNull().deepEquals(expectedTopics);
+      checkNotified(count: messages.length);
+    }
+
+    test('ignore content-only update', () async {
+      final message = eg.streamMessage(id: 123, stream: channel, topic: 'foo');
+      await prepareWithMessages([message], expectedTopics: [
+        isChannelTopicsEntry('foo', 123),
+      ]);
+
+      await store.handleEvent(eg.updateMessageEditEvent(message));
+      checkNotNotified();
+    });
+
+    group('PropagateMode.changedAll', () {
+      test('topic moved to another channel with no previously fetched topics', () async {
+        final messagesToMove = List.generate(10, (i) =>
+          eg.streamMessage(id: 100 + i, stream: channel, topic: 'foo'));
+        await prepareWithMessages(messagesToMove, expectedTopics: [
+          isChannelTopicsEntry('foo', 109),
+        ]);
+
+        await store.handleEvent(eg.updateMessageEventMoveFrom(
+          origMessages: messagesToMove,
+          newStreamId: otherChannel.streamId,
+          propagateMode: .changeAll));
+        check(model).getChannelTopics(channel.streamId).isNotNull().isEmpty();
+        check(model).getChannelTopics(otherChannel.streamId).isNull();
+        checkNotifiedOnce();
+      });
+
+      test('topic moved to new topic in another channel', () async {
+        final messagesToMove = List.generate(10, (i) =>
+          eg.streamMessage(id: 100 + i, stream: channel, topic: 'foo'));
+        await prepareWithMessages(messagesToMove, expectedTopics: [
+          isChannelTopicsEntry('foo', 109),
+        ]);
+
+        // Make sure that topics in otherChannel have been fetched.
+        connection.prepare(json: GetChannelTopicsResult(topics: [
+          eg.getChannelTopicsEntry(name: 'foo', maxId: 1),
+        ]).toJson());
+        await model.fetchChannelTopics(otherChannel.streamId);
+        check(model).getChannelTopics(otherChannel.streamId).isNotNull()
+          .single.which(isChannelTopicsEntry('foo', 1));
+        checkNotNotified();
+
+        await store.handleEvent(eg.updateMessageEventMoveFrom(
+          origMessages: messagesToMove,
+          newStreamId: otherChannel.streamId,
+          newTopicStr: 'bar',
+          propagateMode: .changeAll));
+        check(model).getChannelTopics(channel.streamId).isNotNull().isEmpty();
+        check(model).getChannelTopics(otherChannel.streamId).isNotNull().deepEquals([
+          isChannelTopicsEntry('bar', 109),
+          isChannelTopicsEntry('foo', 1),
+        ]);
+        checkNotifiedOnce();
+      });
+
+      test('topic moved to known topic in another channel', () async {
```

Comment:
I think "existing topic" would be clearer than "known topic". It's true that a topic being "unknown" (because of [protected history](https://zulip.com/help/channel-permissions#private-channels)) can be the reason it's absent in our data structures. But that feels like an edge case that this test doesn't need to focus on, and the more common/simpler scenario is just that the topic doesn't exist yet (because no messages have been sent in it yet).